packages/modules/pricing/src/repositories/pricing.ts

```diff
@@ -95,198 +80,197 @@ export class PricingRepository
-            ">=",
-            knex.fn.now()
-          )
+        .andOn(function (this: Knex.JoinClause) {
+          this.onNull("pl.ends_at").orOn("pl.ends_at", ">=", knex.fn.now())
         })
-        .andWhere(function () {
-          this.andWhere(function () {
-            for (const [key, value] of flattenedContext) {
-              this.orWhere({ "plr.attribute": key })
-              this.where(
-                "plr.value",
-                "@>",
-                JSON.stringify(Array.isArray(value) ? value : [value])
-              )
-            }
-
-            this.orWhere("pl.rules_count", "=", 0)
-          })
-
-          this.andWhere(function () {
-            this.andWhere((contextBuilder) => {
-              for (const [key, value] of flattenedContext) {
-                contextBuilder.orWhere((orBuilder) => {
-                  orBuilder.where("pr.attribute", key)
+    })
 
+    // OPTIMIZATION: Only add complex rule filtering when necessary
+    if (hasComplexContext) {
+      // For price rules - direct check that ALL rules match
+      const priceRuleConditions = knex.raw(
+        `
+        (
+          price.rules_count = 0 OR
+          (
+            /* Count all matching rules and compare to total rule count */
+            SELECT COUNT(*) 
+            FROM price_rule pr
+            WHERE pr.price_id = price.id 
+            AND pr.deleted_at IS NULL
+            AND (
+              ${flattenedContext
+                .map(([key, value]) => {
                   if (typeof value === "number") {
-                    buildOperatorQueries(orBuilder, value)
+                    return `
+                    (pr.attribute = ? AND (
+                      (pr.operator = 'eq' AND pr.value::numeric = ?) OR
+                      (pr.operator = 'gt' AND ? > pr.value::numeric) OR
+                      (pr.operator = 'gte' AND ? >= pr.value::numeric) OR
+                      (pr.operator = 'lt' AND ? < pr.value::numeric) OR
+                      (pr.operator = 'lte' AND ? <= pr.value::numeric)
+                    ))
+                    `
                   } else {
                     const normalizeValue = Array.isArray(value)
                       ? value
                       : [value]
-
-                    orBuilder.whereIn("pr.value", normalizeValue)
+                    const placeholders = normalizeValue.map(() => "?").join(",")
+                    return `(pr.attribute = ? AND pr.value IN (${placeholders}))`
                   }
                 })
-              }
+                .join(" OR ")}
+            )
+          ) = (
+            /* Get total rule count */
+            SELECT COUNT(*) 
+            FROM price_rule pr
+            WHERE pr.price_id = price.id 
+            AND pr.deleted_at IS NULL
+          )
+        )
+        `,
+        flattenedContext.flatMap(([key, value]) => {
+          if (typeof value === "number") {
+            return [key, value, value, value, value, value]
```

Comment:
This will fix a bug we have right now

```suggestion
            return [key, value.toString(), value, value, value, value]
```